File: ps1/mergesort/main.cpp
#include <string>
#include <iostream>
#include <fstream>
#include <cstdlib>
#include <vector>

using std::string;
using std::vector;
using std::ifstream;
using std::getline;
using std::stoi;
using std::cout;
using std::endl;
using std::distance;
using std::copy;

namespace {
vector<int> get_data(const string& filename) {
  ifstream file(filename.c_str());
  if(!file.good())
    return vector<int>();

  string line;
  getline(file, line); // get the header
  getline(file, line); // get the count
  const int len = stoi(line);
  vector<int> data(len);
  for(int i = 0; getline(file, line); ++i) {
    data[i] = stoi(line);
  }
  return data;
}

inline void print_data(const vector<int>& data) {
  cout << "[";
  for(vector<int>::const_iterator i = data.begin(), end = data.end(); i != end; ++i)
    cout << *i << " ";
  cout << "]";
}

inline void print_data(vector<int>::const_iterator first, vector<int>::const_iterator last) {
  cout << "[";
  for(; first != last; ++first)
    cout << *first << " ";
  cout << "]";
}

void mergesort(vector<int>::iterator first, vector<int>::iterator last) {
  auto swap = [](int* a, int* b) {
    const auto c = *a;
    *a = *b;
    *b = c;
  };

  auto merge = [](vector<int>::const_iterator abegin, vector<int>::const_iterator aend,
                  vector<int>::const_iterator bbegin, vector<int>::const_iterator bend) {
    vector<int> buffer(distance(abegin, aend) + distance(bbegin, bend));
    auto bufferi = buffer.begin();
    auto ai = abegin;
    auto bi = bbegin;
    for(; ai != aend && bi != bend; ++bufferi) {
      if(*ai < *bi) {
        *bufferi = *ai;
        ++ai;
      } else {
        *bufferi = *bi;
        ++bi;
      }
    }

    if(ai != aend)
      copy(ai, aend, bufferi);
    else if(bi != bend)
      copy(bi, bend, bufferi);

    return buffer;
  };

  const auto len = distance(first, last);

  if(len > 2) {
    const auto split = len / 2;
    mergesort(first, first + split);
    mergesort(first + split, last);
    const auto buffer = merge(first, first + split, first + split, last);
    copy(buffer.begin(), buffer.end(), first);
    return;
  }

  if(len == 2) {
    if(*first > *(first + 1))
      swap(&(*first), &(*(first + 1)));
    return;
  }

  return;
}

} // namespace

int main(int argc, char* argv[]) {
  if(argc < 2) {
    cout << "usage: " << argv[0] << " inputfile" << endl;
    return 0;
  }

  const string filename = argv[1];
  vector<int> data = get_data(filename);
  if(data.size() == 0) {
    cout << "failed to get data" << endl;
    return 0;
  }

  print_data(data);
  cout << endl;
  cout << "sorting..." << endl;
  mergesort(data.begin(), data.end());

  print_data(data);
  cout << endl;

  return 0;
}

File: ps1/mergesort/main.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"os"
	"strconv"
)

var inputfile = flag.String("i", "", "The data input file to read from")

func getdata(filename string) ([]int, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer func() {
		file.Close() /// \todo handle error
	}()

	scanner := bufio.NewScanner(file)
	scanner.Scan() // scan the header line
	scanner.Scan() // scan the length of the file
	len, err := strconv.Atoi(scanner.Text())
	if err != nil {
		return nil, err
	}

	data := make([]int, len, len)

	for i := 0; scanner.Scan(); i++ {
		val, err := strconv.Atoi(scanner.Text())
		if err != nil {
			return nil, err
		}
		data[i] = val
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}

	return data, nil
}

func mergesort(data []int) []int {
	/// DO NOT replace this with the XOR method. It is slow.
	swap := func(a *int, b *int) {
		temp := *a
		*a = *b
		*b = temp
	}

	merge := func(a []int, b []int) []int {
		buffer := make([]int, len(a)+len(b), len(a)+len(b))
		ai, bi := 0, 0
		for ai < len(a) && bi < len(b) {
			if a[ai] < b[bi] {
				buffer[ai+bi] = a[ai]
				ai++
			} else {
				buffer[ai+bi] = b[bi]
				bi++
			}
		}

		if ai != len(a) {
			copy(buffer[ai+bi:], a[ai:])
		} else if bi != len(b) {
			copy(buffer[ai+bi:], b[bi:])
		}

		return buffer
	}

	if len(data) > 2 {
		split := len(data) / 2
		/// These recursive mergesorts can be safely, trivially run in parallel with goroutines, for log n speedup for T∞ = log n processors.
		left := mergesort(data[:split])
		right := mergesort(data[split:])
		return merge(left, right) /// Less trivially, the merge may be parallelised for greater parallelism which scales with manycore architectures.
	}

	if len(data) == 2 {
		if data[0] > data[1] {
			swap(&data[0], &data[1])
		}
		return data
	}

	return data
}

func main() {
	flag.Parse()

	if *inputfile == "" {
		fmt.Println("Usage:")
		flag.PrintDefaults()
		return
	}

	data, err := getdata(*inputfile)
	if err != nil {
		fmt.Println(err)
		return
	}

	fmt.Println(data)

	fmt.Println("sorting...")
	data = mergesort(data)

	fmt.Println(data)
}
